File: app/api/mqtt_settings.py
"""
MQTT Settings Management API
Handles MQTT broker configuration, authentication, and connection testing
"""

import os
import json
import logging
import paho.mqtt.client as mqtt
from flask import Blueprint, request, jsonify, current_app
from datetime import datetime
import ssl
import time
from threading import Thread, Event

logger = logging.getLogger(__name__)

bp = Blueprint('mqtt_settings', __name__)

# Global MQTT client for testing connections
test_client = None
test_result = {}

class MQTTConfig:
    """MQTT Configuration Manager"""
    
    DEFAULT_CONFIG = {
        # Connection settings
        'host': 'localhost',
        'port': 1883,
        'use_ssl': False,
        'ssl_port': 8883,
        'ssl_cert_path': '',
        'ssl_key_path': '',
        'ssl_ca_path': '',
        'ssl_insecure': False,
        
        # Authentication
        'use_auth': False,
        'username': '',
        'password': '',
        'client_id': 'erik-image-manager',
        
        # Connection parameters
        'keepalive': 60,
        'connect_timeout': 30,
        'reconnect_on_failure': True,
        'reconnect_delay': 5,
        'max_reconnect_attempts': 10,
        
        # Topics
        'topic_prefix': 'frigate',
        'custom_topics': {
            'events': '{prefix}/events',
            'detection': '{prefix}/+/person',
            'tracking': 'erik/tracking/+',
            'pose': 'erik/pose/+',
            'commands': 'erik/commands/+'
        },
        'qos_level': 1,  # 0: At most once, 1: At least once, 2: Exactly once
        
        # Advanced settings
        'clean_session': True,
        'protocol_version': 4,  # 3: MQTTv3.1, 4: MQTTv3.1.1, 5: MQTTv5
        'transport': 'tcp',  # tcp or websockets
        
        # Last Will and Testament
        'use_lwt': False,
        'lwt_topic': 'erik/status',
        'lwt_message': 'offline',
        'lwt_qos': 1,
        'lwt_retain': True,
        
        # Message handling
        'retain_messages': False,
        'message_buffer_size': 1000,
        
        # Status and debugging
        'enable_logging': True,
        'log_level': 'INFO',
        'enable_metrics': False,
        'metrics_interval': 60
    }
    
    def __init__(self, config_path='/home/andrew/nvr/config/mqtt_config.json'):
        self.config_path = config_path
        self.config = self.load_config()
        self.connection_status = {
            'connected': False,
            'last_connection': None,
            'last_error': None,
            'messages_received': 0,
            'messages_sent': 0,
            'uptime': 0
        }
    
    def load_config(self):
        """Load MQTT configuration from file"""
        if os.path.exists(self.config_path):
            try:
                with open(self.config_path, 'r') as f:
                    saved_config = json.load(f)
                    # Merge with defaults to ensure all keys exist
                    config = self.DEFAULT_CONFIG.copy()
                    config.update(saved_config)
                    return config
            except Exception as e:
                logger.error(f"Error loading MQTT config: {e}")
        return self.DEFAULT_CONFIG.copy()
    
    def save_config(self, config):
        """Save MQTT configuration to file"""
        try:
            os.makedirs(os.path.dirname(self.config_path), exist_ok=True)
            with open(self.config_path, 'w') as f:
                json.dump(config, f, indent=2)
            self.config = config
            return True
        except Exception as e:
            logger.error(f"Error saving MQTT config: {e}")
            return False
    
    def validate_config(self, config):
        """Validate MQTT configuration"""
        errors = []
        
        # Validate host
        if not config.get('host'):
            errors.append('Host is required')
        
        # Validate port
        port = config.get('port')
        if not isinstance(port, int) or port < 1 or port > 65535:
            errors.append('Port must be between 1 and 65535')
        
        # Validate SSL settings
        if config.get('use_ssl'):
            if config.get('ssl_cert_path') and not os.path.exists(config['ssl_cert_path']):
                errors.append(f"SSL certificate not found: {config['ssl_cert_path']}")
            if config.get('ssl_key_path') and not os.path.exists(config['ssl_key_path']):
                errors.append(f"SSL key not found: {config['ssl_key_path']}")
            if config.get('ssl_ca_path') and not os.path.exists(config['ssl_ca_path']):
                errors.append(f"SSL CA certificate not found: {config['ssl_ca_path']}")
        
        # Validate authentication
        if config.get('use_auth'):
            if not config.get('username'):
                errors.append('Username is required when authentication is enabled')
        
        # Validate QoS level
        qos = config.get('qos_level', 1)
        if qos not in [0, 1, 2]:
            errors.append('QoS level must be 0, 1, or 2')
        
        # Validate protocol version
        protocol = config.get('protocol_version', 4)
        if protocol not in [3, 4, 5]:
            errors.append('Protocol version must be 3 (MQTTv3.1), 4 (MQTTv3.1.1), or 5 (MQTTv5)')
        
        return errors

# Create global config manager
mqtt_config_manager = MQTTConfig()

@bp.route('/config', methods=['GET'])
def get_mqtt_config():
    """Get current MQTT configuration"""
    try:
        config = mqtt_config_manager.config
        status = mqtt_config_manager.connection_status
        
        return jsonify({
            'success': True,
            'config': config,
            'status': status
        })
    except Exception as e:
        logger.error(f"Error getting MQTT config: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/config', methods=['POST'])
def update_mqtt_config():
    """Update MQTT configuration"""
    try:
        data = request.json
        
        # Validate configuration
        errors = mqtt_config_manager.validate_config(data)
        if errors:
            return jsonify({
                'success': False,
                'errors': errors
            }), 400
        
        # Save configuration
        if mqtt_config_manager.save_config(data):
            logger.info("MQTT configuration updated successfully")
            
            # Restart MQTT listener if it's running
            # This would trigger a reconnection with new settings
            if hasattr(current_app, 'mqtt_listener'):
                logger.info("Restarting MQTT listener with new configuration")
                # Implementation would restart the listener
            
            return jsonify({
                'success': True,
                'message': 'MQTT configuration updated successfully'
            })
        else:
            return jsonify({
                'success': False,
                'error': 'Failed to save configuration'
            }), 500
            
    except Exception as e:
        logger.error(f"Error updating MQTT config: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/test', methods=['POST'])
def test_mqtt_connection():
    """Test MQTT connection with current or provided settings"""
    global test_client, test_result
    
    try:
        # Get test configuration (either from request or current config)
        test_config = request.json if request.json else mqtt_config_manager.config
        
        # Reset test result
        test_result = {
            'connecting': True,
            'connected': False,
            'error': None,
            'details': {}
        }
        
        # Create test client
        client_id = f"{test_config.get('client_id', 'test')}-{int(time.time())}"
        # Get protocol version
        protocol_version = test_config.get('protocol_version', 4)
        if protocol_version == 3:
            protocol = mqtt.MQTTv31
        elif protocol_version == 5:
            protocol = mqtt.MQTTv5
        else:
            protocol = mqtt.MQTTv311  # Default to 3.1.1
        
        test_client = mqtt.Client(
            client_id=client_id,
            clean_session=test_config.get('clean_session', True),
            protocol=protocol
        )
        
        # Setup callbacks
        def on_connect(client, userdata, flags, rc):
            if rc == 0:
                test_result['connected'] = True
                test_result['details'] = {
                    'broker': test_config['host'],
                    'port': test_config['port'],
                    'client_id': client_id,
                    'protocol': f"MQTTv{test_config.get('protocol_version', '3.1.1')}"
                }
                logger.info(f"MQTT test connection successful to {test_config['host']}:{test_config['port']}")
            else:
                error_messages = {
                    1: "Incorrect protocol version",
                    2: "Invalid client identifier",
                    3: "Server unavailable",
                    4: "Bad username or password",
                    5: "Not authorized"
                }
                test_result['error'] = error_messages.get(rc, f"Connection failed with code {rc}")
                logger.error(f"MQTT test connection failed: {test_result['error']}")
            test_result['connecting'] = False
            client.disconnect()
        
        def on_disconnect(client, userdata, rc):
            test_result['connecting'] = False
        
        test_client.on_connect = on_connect
        test_client.on_disconnect = on_disconnect
        
        # Configure authentication
        if test_config.get('use_auth'):
            test_client.username_pw_set(
                test_config.get('username'),
                test_config.get('password')
            )
        
        # Configure SSL/TLS
        if test_config.get('use_ssl'):
            context = ssl.create_default_context()
            
            if test_config.get('ssl_ca_path'):
                context.load_verify_locations(test_config['ssl_ca_path'])
            
            if test_config.get('ssl_cert_path') and test_config.get('ssl_key_path'):
                context.load_cert_chain(
                    test_config['ssl_cert_path'],
                    test_config['ssl_key_path']
                )
            
            if test_config.get('ssl_insecure'):
                context.check_hostname = False
                context.verify_mode = ssl.CERT_NONE
            
            test_client.tls_set_context(context)
        
        # Configure Last Will and Testament
        if test_config.get('use_lwt'):
            test_client.will_set(
                test_config.get('lwt_topic', 'test/status'),
                test_config.get('lwt_message', 'offline'),
                qos=test_config.get('lwt_qos', 1),
                retain=test_config.get('lwt_retain', True)
            )
        
        # Attempt connection
        port = test_config.get('ssl_port' if test_config.get('use_ssl') else 'port', 1883)
        test_client.connect(
            test_config['host'],
            port,
            test_config.get('keepalive', 60)
        )
        
        # Run in separate thread to avoid blocking
        def run_test():
            test_client.loop_start()
            # Wait for connection result (max 10 seconds)
            timeout = 10
            start_time = time.time()
            while test_result['connecting'] and (time.time() - start_time) < timeout:
                time.sleep(0.1)
            test_client.loop_stop()
            
            if test_result['connecting']:
                test_result['connecting'] = False
                test_result['error'] = f"Connection timeout after {timeout} seconds"
        
        test_thread = Thread(target=run_test)
        test_thread.daemon = True
        test_thread.start()
        
        # Wait a moment for the connection to establish
        time.sleep(0.5)
        
        # Return immediate response (test continues in background)
        return jsonify({
            'success': True,
            'message': 'Connection test initiated',
            'test_id': client_id
        })
        
    except Exception as e:
        logger.error(f"Error testing MQTT connection: {e}")
        return jsonify({
            'success': False,
            'error': str(e)
        }), 500

@bp.route('/test/status', methods=['GET'])
def get_test_status():
    """Get status of ongoing connection test"""
    global test_result
    
    return jsonify({
        'success': True,
        'result': test_result
    })

@bp.route('/topics', methods=['GET'])
def get_mqtt_topics():
    """Get configured MQTT topics"""
    try:
        config = mqtt_config_manager.config
        prefix = config.get('topic_prefix', 'frigate')
        
        # Expand topic patterns with prefix
        topics = {}
        for name, pattern in config.get('custom_topics', {}).items():
            topics[name] = pattern.replace('{prefix}', prefix)
        
        return jsonify({
            'success': True,
            'prefix': prefix,
            'topics': topics,
            'qos_level': config.get('qos_level', 1)
        })
        
    except Exception as e:
        logger.error(f"Error getting MQTT topics: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/topics', methods=['POST'])
def update_mqtt_topics():
    """Update MQTT topic configuration"""
    try:
        data = request.json
        config = mqtt_config_manager.config
        
        if 'prefix' in data:
            config['topic_prefix'] = data['prefix']
        
        if 'custom_topics' in data:
            config['custom_topics'].update(data['custom_topics'])
        
        if 'qos_level' in data:
            qos = data['qos_level']
            if qos not in [0, 1, 2]:
                return jsonify({
                    'success': False,
                    'error': 'QoS level must be 0, 1, or 2'
                }), 400
            config['qos_level'] = qos
        
        # Save updated configuration
        if mqtt_config_manager.save_config(config):
            return jsonify({
                'success': True,
                'message': 'MQTT topics updated successfully'
            })
        else:
            return jsonify({
                'success': False,
                'error': 'Failed to save configuration'
            }), 500
            
    except Exception as e:
        logger.error(f"Error updating MQTT topics: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/status', methods=['GET'])
def get_mqtt_status():
    """Get current MQTT connection status and metrics"""
    try:
        # This would integrate with the actual MQTT listener
        # For now, return the stored status
        status = mqtt_config_manager.connection_status
        
        # Add current configuration info
        config = mqtt_config_manager.config
        status['broker'] = f"{config['host']}:{config['port']}"
        status['ssl_enabled'] = config.get('use_ssl', False)
        status['auth_enabled'] = config.get('use_auth', False)
        
        return jsonify({
            'success': True,
            'status': status
        })
        
    except Exception as e:
        logger.error(f"Error getting MQTT status: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

@bp.route('/logs', methods=['GET'])
def get_mqtt_logs():
    """Get recent MQTT activity logs"""
    try:
        # This would retrieve actual MQTT logs
        # For now, return sample data
        logs = [
            {
                'timestamp': datetime.now().isoformat(),
                'level': 'INFO',
                'message': 'MQTT settings module initialized'
            }
        ]
        
        return jsonify({
            'success': True,
            'logs': logs
        })
        
    except Exception as e:
        logger.error(f"Error getting MQTT logs: {e}")
        return jsonify({'success': False, 'error': str(e)}), 500

File: app/static/js/mqtt-settings.js
/**
 * MQTT Settings Manager
 * Handles MQTT broker configuration UI and testing
 */

class MQTTSettingsManager {
    constructor(apiClient) {
        this.api = apiClient;
        this.currentConfig = null;
        this.testInProgress = false;
        this.statusUpdateInterval = null;
        
        // Bind methods
        this.initialize = this.initialize.bind(this);
        this.loadSettings = this.loadSettings.bind(this);
        this.saveSettings = this.saveSettings.bind(this);
        this.testConnection = this.testConnection.bind(this);
    }
    
    /**
     * Initialize MQTT settings UI
     */
    async initialize() {
        await this.loadSettings();
        this.setupEventListeners();
        this.startStatusMonitoring();
    }
    
    /**
     * Load current MQTT settings
     */
    async loadSettings() {
        try {
            const response = await this.api.get('/api/mqtt/config');
            if (response.success) {
                this.currentConfig = response.config;
                this.updateUI(response.config);
                this.updateStatus(response.status);
            }
        } catch (error) {
            console.error('Failed to load MQTT settings:', error);
            Utils.showToast('Failed to load MQTT settings', 'error');
        }
    }
    
    /**
     * Update UI with configuration values
     */
    updateUI(config) {
        // Connection settings
        Utils.safeSetValue('mqtt-host', config.host);
        Utils.safeSetValue('mqtt-port', config.port);
        Utils.safeSetChecked('mqtt-use-ssl', config.use_ssl);
        Utils.safeSetValue('mqtt-ssl-port', config.ssl_port);
        Utils.safeSetValue('mqtt-ssl-cert', config.ssl_cert_path);
        Utils.safeSetValue('mqtt-ssl-key', config.ssl_key_path);
        Utils.safeSetValue('mqtt-ssl-ca', config.ssl_ca_path);
        Utils.safeSetChecked('mqtt-ssl-insecure', config.ssl_insecure);
        
        // Authentication
        Utils.safeSetChecked('mqtt-use-auth', config.use_auth);
        Utils.safeSetValue('mqtt-username', config.username);
        Utils.safeSetValue('mqtt-password', config.password);
        Utils.safeSetValue('mqtt-client-id', config.client_id);
        
        // Connection parameters
        Utils.safeSetValue('mqtt-keepalive', config.keepalive);
        Utils.safeSetValue('mqtt-timeout', config.connect_timeout);
        Utils.safeSetChecked('mqtt-reconnect', config.reconnect_on_failure);
        Utils.safeSetValue('mqtt-reconnect-delay', config.reconnect_delay);
        Utils.safeSetValue('mqtt-max-reconnect', config.max_reconnect_attempts);
        
        // Topics
        Utils.safeSetValue('mqtt-topic-prefix', config.topic_prefix);
        Utils.safeSetValue('mqtt-qos', config.qos_level);
        
        // Custom topics
        if (config.custom_topics) {
            this.updateCustomTopics(config.custom_topics);
        }
        
        // Advanced settings
        Utils.safeSetChecked('mqtt-clean-session', config.clean_session);
        Utils.safeSetValue('mqtt-protocol', config.protocol_version);
        Utils.safeSetValue('mqtt-transport', config.transport);
        
        // Last Will and Testament
        Utils.safeSetChecked('mqtt-use-lwt', config.use_lwt);
        Utils.safeSetValue('mqtt-lwt-topic', config.lwt_topic);
        Utils.safeSetValue('mqtt-lwt-message', config.lwt_message);
        Utils.safeSetValue('mqtt-lwt-qos', config.lwt_qos);
        Utils.safeSetChecked('mqtt-lwt-retain', config.lwt_retain);
        
        // Message handling
        Utils.safeSetChecked('mqtt-retain', config.retain_messages);
        Utils.safeSetValue('mqtt-buffer-size', config.message_buffer_size);
        
        // Debugging
        Utils.safeSetChecked('mqtt-enable-logging', config.enable_logging);
        Utils.safeSetValue('mqtt-log-level', config.log_level);
        Utils.safeSetChecked('mqtt-enable-metrics', config.enable_metrics);
        
        // Update UI state based on toggles
        this.toggleSSLSettings(config.use_ssl);
        this.toggleAuthSettings(config.use_auth);
        this.toggleLWTSettings(config.use_lwt);
    }
    
    /**
     * Update custom topics display
     */
    updateCustomTopics(topics) {
        const container = document.getElementById('mqtt-custom-topics');
        if (!container) return;
        
        let html = '';
        for (const [name, pattern] of Object.entries(topics)) {
            html += `
                <div class="custom-topic-row">
                    <input type="text" class="topic-name" value="${name}" placeholder="Topic name">
                    <input type="text" class="topic-pattern" value="${pattern}" placeholder="Topic pattern">
                    <button class="btn-remove" onclick="mqttSettings.removeTopic('${name}')">×</button>
                </div>
            `;
        }
        container.innerHTML = html;
    }
    
    /**
     * Setup event listeners
     */
    setupEventListeners() {
        // SSL toggle
        Utils.safeElementOperation('mqtt-use-ssl', el => {
            el.addEventListener('change', (e) => this.toggleSSLSettings(e.target.checked));
        });
        
        // Auth toggle
        Utils.safeElementOperation('mqtt-use-auth', el => {
            el.addEventListener('change', (e) => this.toggleAuthSettings(e.target.checked));
        });
        
        // LWT toggle
        Utils.safeElementOperation('mqtt-use-lwt', el => {
            el.addEventListener('change', (e) => this.toggleLWTSettings(e.target.checked));
        });
        
        // Save button
        Utils.safeElementOperation('mqtt-save-btn', el => {
            el.addEventListener('click', () => this.saveSettings());
        });
        
        // Test button
        Utils.safeElementOperation('mqtt-test-btn', el => {
            el.addEventListener('click', () => this.testConnection());
        });
        
        // Add topic button
        Utils.safeElementOperation('mqtt-add-topic-btn', el => {
            el.addEventListener('click', () => this.addCustomTopic());
        });
        
        // Port sync when SSL is toggled
        Utils.safeElementOperation('mqtt-use-ssl', el => {
            el.addEventListener('change', (e) => {
                if (e.target.checked) {
                    const portField = document.getElementById('mqtt-port');
                    const sslPortField = document.getElementById('mqtt-ssl-port');
                    if (portField && portField.value === '1883' && sslPortField) {
                        portField.value = sslPortField.value || '8883';
                    }
                } else {
                    const portField = document.getElementById('mqtt-port');
                    if (portField && portField.value === '8883') {
                        portField.value = '1883';
                    }
                }
            });
        });
    }
    
    /**
     * Toggle SSL settings visibility
     */
    toggleSSLSettings(enabled) {
        const sslSettings = document.getElementById('mqtt-ssl-settings');
        if (sslSettings) {
            sslSettings.style.display = enabled ? 'block' : 'none';
        }
    }
    
    /**
     * Toggle authentication settings visibility
     */
    toggleAuthSettings(enabled) {
        const authSettings = document.getElementById('mqtt-auth-settings');
        if (authSettings) {
            authSettings.style.display = enabled ? 'block' : 'none';
        }
    }
    
    /**
     * Toggle LWT settings visibility
     */
    toggleLWTSettings(enabled) {
        const lwtSettings = document.getElementById('mqtt-lwt-settings');
        if (lwtSettings) {
            lwtSettings.style.display = enabled ? 'block' : 'none';
        }
    }
    
    /**
     * Gather configuration from UI
     */
    gatherConfig() {
        const config = {
            // Connection settings
            host: document.getElementById('mqtt-host')?.value || 'localhost',
            port: parseInt(document.getElementById('mqtt-port')?.value || '1883'),
            use_ssl: document.getElementById('mqtt-use-ssl')?.checked || false,
            ssl_port: parseInt(document.getElementById('mqtt-ssl-port')?.value || '8883'),
            ssl_cert_path: document.getElementById('mqtt-ssl-cert')?.value || '',
            ssl_key_path: document.getElementById('mqtt-ssl-key')?.value || '',
            ssl_ca_path: document.getElementById('mqtt-ssl-ca')?.value || '',
            ssl_insecure: document.getElementById('mqtt-ssl-insecure')?.checked || false,
            
            // Authentication
            use_auth: document.getElementById('mqtt-use-auth')?.checked || false,
            username: document.getElementById('mqtt-username')?.value || '',
            password: document.getElementById('mqtt-password')?.value || '',
            client_id: document.getElementById('mqtt-client-id')?.value || 'erik-image-manager',
            
            // Connection parameters
            keepalive: parseInt(document.getElementById('mqtt-keepalive')?.value || '60'),
            connect_timeout: parseInt(document.getElementById('mqtt-timeout')?.value || '30'),
            reconnect_on_failure: document.getElementById('mqtt-reconnect')?.checked !== false,
            reconnect_delay: parseInt(document.getElementById('mqtt-reconnect-delay')?.value || '5'),
            max_reconnect_attempts: parseInt(document.getElementById('mqtt-max-reconnect')?.value || '10'),
            
            // Topics
            topic_prefix: document.getElementById('mqtt-topic-prefix')?.value || 'frigate',
            qos_level: parseInt(document.getElementById('mqtt-qos')?.value || '1'),
            custom_topics: this.gatherCustomTopics(),
            
            // Advanced settings
            clean_session: document.getElementById('mqtt-clean-session')?.checked !== false,
            protocol_version: parseInt(document.getElementById('mqtt-protocol')?.value || '4'),
            transport: document.getElementById('mqtt-transport')?.value || 'tcp',
            
            // Last Will and Testament
            use_lwt: document.getElementById('mqtt-use-lwt')?.checked || false,
            lwt_topic: document.getElementById('mqtt-lwt-topic')?.value || 'erik/status',
            lwt_message: document.getElementById('mqtt-lwt-message')?.value || 'offline',
            lwt_qos: parseInt(document.getElementById('mqtt-lwt-qos')?.value || '1'),
            lwt_retain: document.getElementById('mqtt-lwt-retain')?.checked !== false,
            
            // Message handling
            retain_messages: document.getElementById('mqtt-retain')?.checked || false,
            message_buffer_size: parseInt(document.getElementById('mqtt-buffer-size')?.value || '1000'),
            
            // Debugging
            enable_logging: document.getElementById('mqtt-enable-logging')?.checked !== false,
            log_level: document.getElementById('mqtt-log-level')?.value || 'INFO',
            enable_metrics: document.getElementById('mqtt-enable-metrics')?.checked || false,
            metrics_interval: 60
        };
        
        return config;
    }
    
    /**
     * Gather custom topics from UI
     */
    gatherCustomTopics() {
        const topics = {};
        const rows = document.querySelectorAll('#mqtt-custom-topics .custom-topic-row');
        
        rows.forEach(row => {
            const name = row.querySelector('.topic-name')?.value;
            const pattern = row.querySelector('.topic-pattern')?.value;
            if (name && pattern) {
                topics[name] = pattern;
            }
        });
        
        // Include default topics if not already present
        if (!topics.events) topics.events = '{prefix}/events';
        if (!topics.detection) topics.detection = '{prefix}/+/person';
        if (!topics.tracking) topics.tracking = 'erik/tracking/+';
        
        return topics;
    }
    
    /**
     * Save MQTT settings
     */
    async saveSettings() {
        try {
            const config = this.gatherConfig();
            
            // Show saving indicator
            const saveBtn = document.getElementById('mqtt-save-btn');
            const originalText = saveBtn.textContent;
            saveBtn.textContent = 'Saving...';
            saveBtn.disabled = true;
            
            const response = await this.api.post('/api/mqtt/config', config);
            
            if (response.success) {
                Utils.showToast('MQTT settings saved successfully', 'success');
                this.currentConfig = config;
                
                // Restart connection with new settings
                if (document.getElementById('mqtt-auto-restart')?.checked) {
                    Utils.showToast('Restarting MQTT connection...', 'info');
                }
            } else {
                const errors = response.errors || [response.error];
                Utils.showToast(`Failed to save: ${errors.join(', ')}`, 'error');
            }
            
            saveBtn.textContent = originalText;
            saveBtn.disabled = false;
            
        } catch (error) {
            console.error('Failed to save MQTT settings:', error);
            Utils.showToast('Failed to save MQTT settings', 'error');
            
            const saveBtn = document.getElementById('mqtt-save-btn');
            saveBtn.textContent = 'Save Settings';
            saveBtn.disabled = false;
        }
    }
    
    /**
     * Test MQTT connection
     */
    async testConnection() {
        if (this.testInProgress) {
            Utils.showToast('Test already in progress', 'warning');
            return;
        }
        
        try {
            this.testInProgress = true;
            const config = this.gatherConfig();
            
            // Update UI
            const testBtn = document.getElementById('mqtt-test-btn');
            const originalText = testBtn.textContent;
            testBtn.textContent = 'Testing...';
            testBtn.disabled = true;
            
            const statusDiv = document.getElementById('mqtt-test-status');
            if (statusDiv) {
                statusDiv.innerHTML = '<div class="test-progress">🔄 Connecting to broker...</div>';
                statusDiv.style.display = 'block';
            }
            
            // Start test
            const response = await this.api.post('/api/mqtt/test', config);
            
            if (response.success) {
                // Poll for test result
                this.pollTestStatus(response.test_id, testBtn, originalText, statusDiv);
            } else {
                throw new Error(response.error || 'Test failed to start');
            }
            
        } catch (error) {
            console.error('Failed to test connection:', error);
            Utils.showToast(`Connection test failed: ${error.message}`, 'error');
            
            const testBtn = document.getElementById('mqtt-test-btn');
            testBtn.textContent = 'Test Connection';
            testBtn.disabled = false;
            this.testInProgress = false;
            
            const statusDiv = document.getElementById('mqtt-test-status');
            if (statusDiv) {
                statusDiv.innerHTML = `<div class="test-error">❌ ${error.message}</div>`;
            }
        }
    }
    
    /**
     * Poll for test connection status
     */
    async pollTestStatus(testId, button, originalText, statusDiv) {
        let attempts = 0;
        const maxAttempts = 20; // 10 seconds max
        
        const checkStatus = async () => {
            try {
                const response = await this.api.get('/api/mqtt/test/status');
                
                if (!response.result.connecting) {
                    // Test complete
                    button.textContent = originalText;
                    button.disabled = false;
                    this.testInProgress = false;
                    
                    if (response.result.connected) {
                        Utils.showToast('Connection successful!', 'success');
                        if (statusDiv) {
                            const details = response.result.details;
                            statusDiv.innerHTML = `
                                <div class="test-success">
                                    ✅ Connected successfully!
                                    <div class="test-details">
                                        <div>Broker: ${details.broker}</div>
                                        <div>Port: ${details.port}</div>
                                        <div>Client ID: ${details.client_id}</div>
                                        <div>Protocol: ${details.protocol}</div>
                                    </div>
                                </div>
                            `;
                        }
                    } else {
                        const error = response.result.error || 'Connection failed';
                        Utils.showToast(error, 'error');
                        if (statusDiv) {
                            statusDiv.innerHTML = `<div class="test-error">❌ ${error}</div>`;
                        }
                    }
                    return;
                }
                
                // Still connecting
                attempts++;
                if (attempts >= maxAttempts) {
                    throw new Error('Connection test timeout');
                }
                
                // Continue polling
                setTimeout(checkStatus, 500);
                
            } catch (error) {
                console.error('Error polling test status:', error);
                button.textContent = originalText;
                button.disabled = false;
                this.testInProgress = false;
                
                if (statusDiv) {
                    statusDiv.innerHTML = `<div class="test-error">❌ ${error.message}</div>`;
                }
            }
        };
        
        // Start polling
        checkStatus();
    }
    
    /**
     * Add custom topic
     */
    addCustomTopic() {
        const container = document.getElementById('mqtt-custom-topics');
        if (!container) return;
        
        const row = document.createElement('div');
        row.className = 'custom-topic-row';
        row.innerHTML = `
            <input type="text" class="topic-name" placeholder="Topic name">
            <input type="text" class="topic-pattern" placeholder="Topic pattern (use {prefix} for prefix)">
            <button class="btn-remove" onclick="this.parentElement.remove()">×</button>
        `;
        container.appendChild(row);
    }
    
    /**
     * Remove custom topic
     */
    removeTopic(name) {
        const rows = document.querySelectorAll('#mqtt-custom-topics .custom-topic-row');
        rows.forEach(row => {
            const nameInput = row.querySelector('.topic-name');
            if (nameInput && nameInput.value === name) {
                row.remove();
            }
        });
    }
    
    /**
     * Update connection status display
     */
    updateStatus(status) {
        // Update status indicators
        Utils.safeElementOperation('mqtt-status-connected', el => {
            el.textContent = status.connected ? '🟢 Connected' : '🔴 Disconnected';
            el.className = status.connected ? 'status-connected' : 'status-disconnected';
        });
        
        Utils.safeElementOperation('mqtt-status-broker', el => {
            el.textContent = status.broker || 'Not configured';
        });
        
        Utils.safeElementOperation('mqtt-status-messages-rx', el => {
            el.textContent = status.messages_received || '0';
        });
        
        Utils.safeElementOperation('mqtt-status-messages-tx', el => {
            el.textContent = status.messages_sent || '0';
        });
        
        if (status.last_error) {
            Utils.safeElementOperation('mqtt-status-error', el => {
                el.textContent = status.last_error;
                el.style.display = 'block';
            });
        }
    }
    
    /**
     * Start monitoring connection status
     */
    startStatusMonitoring() {
        // Update status every 5 seconds
        this.statusUpdateInterval = setInterval(async () => {
            try {
                const response = await this.api.get('/api/mqtt/status');
                if (response.success) {
                    this.updateStatus(response.status);
                }
            } catch (error) {
                console.error('Failed to update MQTT status:', error);
            }
        }, 5000);
    }
    
    /**
     * Stop status monitoring
     */
    stopStatusMonitoring() {
        if (this.statusUpdateInterval) {
            clearInterval(this.statusUpdateInterval);
            this.statusUpdateInterval = null;
        }
    }
    
    /**
     * Load and display MQTT logs
     */
    async loadLogs() {
        try {
            const response = await this.api.get('/api/mqtt/logs');
            if (response.success) {
                this.displayLogs(response.logs);
            }
        } catch (error) {
            console.error('Failed to load MQTT logs:', error);
        }
    }
    
    /**
     * Display MQTT logs
     */
    displayLogs(logs) {
        const container = document.getElementById('mqtt-logs');
        if (!container) return;
        
        let html = '';
        logs.forEach(log => {
            const levelClass = log.level.toLowerCase();
            html += `
                <div class="log-entry log-${levelClass}">
                    <span class="log-time">${new Date(log.timestamp).toLocaleTimeString()}</span>
                    <span class="log-level">[${log.level}]</span>
                    <span class="log-message">${log.message}</span>
                </div>
            `;
        });
        
        container.innerHTML = html || '<div class="no-logs">No logs available</div>';
    }
}

// Create global instance
window.mqttSettings = new MQTTSettingsManager(window.api);